### ISSUE TITLE
EMACROSS from candle 0?

### ISSUE THREAD

**Issue by ITJusus** [taniman/profit-trailer#1376]
There are a few requests on here related EMACROSS, but none of them seem to specifically address the issue I'm running into.  Please correct me if I'm wrong or if this feature exists and there is already a way to do this.

I'm trying to execute a buy just as the current price crosses up the trigger % level (buy_value) of the CURRENT candle (EMA_cross_candles = 0).  In other words, suppose I'm using the following settings...

ALL_buy_strategy = EMACROSS
ALL_buy_value = 0
EMA_cross_candles =  0

My expectation was that setting the cross_candles to 0 would mean "ground 0" (the current candle).  The most recent version of PT (v 1.2.6.21) says that a buy_value of 0 should now correctly trigger a buy on the cross up.  But, these settings are generating 0 buys.  My theory is that setting the cross_candles to 0 is somehow nullifying the strategy.  

Does this strategy not support triggering a buy on the current candle crossing up?  If not, can this feature be added to PT so that EMACROSS does work properly when trying to catch an EMA cross up on the current candle?  Catching the cross 1 or more candles back too often results in a buy entry AFTER the upward move has ended and so is the worst time to enter a buy, in my view.

**Comment by YoJeff** [comment]
You realize that as price unfolds 'in the current candle'  it's potentially flapping above and below until the candle is nailed down?    A cross inside a single candle is rarely clean.

**Comment by ITJusus** [comment]
Yes, I do.  But I think rarely is anything truly clean in these markets...or there would be many more millionaires among us.  It seems to me, though, getting in on the initial cross (candle 0) more often yields positive results than getting in on the next candles.  I know that strategy (entering on later candles) has some level of success in higher volume currencies in the FX market, but much more often in the cryptocurrency market it seems the first candle crossing up the EMA is the largest one and then it just fizzles out from there (especially in longer time-frames).  

In those instances where the price does temporarily go up above the initial EMA Cross point and then sinks back below...that's what we have DCA strategies for.  I've not seen one single strategy out there that guarantees 0 bags and 100% clean profit ratios.  We're all just trying to get our profit to bag ratios as consistently on the profit side as possible.  But ALL of us experience the price dipping down when we thought it was rising.  Just part of the game and then DCA to the rescue (plan B).

**Comment by TheBigShort111** [comment]
I agree completely that we should be able to specify a "zero" candle as a trigger, so long as the fast EMA crosses up the slow, a buy would happen. ITJusus echos my thoughts on the logic and reasons for the cross strategy to work this way, both on the buy and sell side. Waiting a number candles to confirm a cross does not make much sense.

That said, however, as YoJeff pointed out, the cross is often not clean, so perhaps a control or two could be implemented to validate the cross. A couple of suggestions would be a user-specified minimal spread to act as the confirmation trigger. PT already has the ability to see the cross point, and for the user to set a spread. Another control to be used in conjunction with the spread could be a timeout that can be set by the user to substantiate the cross. For example, once a buy is triggered, a sell or rebuy cannot not happen till a user specified amount of time elapses. In this case, the buy happens "very close" to the cross and is committed for a short duration to validate the cross.

Or alternatively, the buy does not happen till the close of the cross candle. In this case, the buy would happen at the close of the cross candle so long as the price remains over the EMA cross point at the candle close. (This method would be the least desirable as often times the cross candle can represent a significant gain) - That said, however, I would still take ANY of the methods above to validate the cross point. From that point, if there is a decline and a cross downward, then a sell will happen, and we would just chalk that up to a swing and a miss.

Below is an image to demonstrate the logic...

Frankly speaking, either way would be better than using EMASPREAD, which rarely buys at the bottom, and often times buys closer to the top during the beginning of a bear runs, resulting in heavy bags.

Thanks again to the devs for your hard work creating an effective EMACROSS sell strategy, in addition to refining EMACROSS as a buy strategy. You are going to make a lot of people very happy!

![screenshot 2018-03-07 16 28 20](https://user-images.githubusercontent.com/36313988/37119880-d405991e-2226-11e8-804c-912e42d8ef6f.png)

**Comment by taniman** [comment]
@TheBigShort111 Please specify how you think this will work.  
You have 2 lines  how do I know they just crossed.

**Comment by TheBigShort111** [comment]
I thought the bot would use the info from the "current value" and compare that to a "buy value" to determine if the fast EMA was above or below the slow EMA. A positive value being above and a negative being below. Wouldn't a zero value indicate the cross point (albeit the cross did not happen yet at zero) - So if the user set a very small positive "buy value" (0.05) that would trigger the buy. No further candles to compare, no trailing... Just use the buy value as the trigger. Same scenario could happen on the sell side with a small negative buy value. In other words, the buy value would become the absolute trigger.

Would/could this not work?

**Comment by ITJusus** [comment]
I would think there would need to be some kind of trailing buy mechanism in place.  The trigger for the start of the trailing buy would be when the bot detects that on the last instant in time when the "bot check" occurred, the Fast EMA was below or equal to the 0 line (Slow EMA)...but now on this current instant in time when the bot check occurs, it detects that the Fast EMA is above the 0 line (Slow EMA).  

This would represent a cross of the EMA's between the last bot check and the current bot check as well as should keep it from entering a trade other than when the Fast EMA has "just crossed up" the Slow EMA.  Once it has crossed, according to this logic, then the buy trail can begin up until it reaches the set buy_value...then buy is triggered when the EMASPREAD is equal or greater than the buy_value.  This is like a normal EMASPREAD is trailed for buy (post initial onset of trailing is triggered), but simply in the opposite direction.

**Comment by TheBigShort111** [comment]
I think what Taniman is eluding to is that each EMA line looks the same to the bot... One is not the EMA fast or slow per se, but rather one line is above the other, almost generically. Perhaps I am wrong in my thinking?

Either way, I think trailing is a bad idea for a cross strategy. What you outlined above is similar to what I have expressed as using a small spread for the trigger. Trailing is a process looking for a lower buy or higher sell, then giving some of that gain up for the trail itself... That does not make much sense for a cross strategy since we are looking to buy as low to the low cross as possible, then sell as close to the high cross as possible. So, for the sake of a trigger, what you refer to as trailing in your last post is more like "monitoring" (or tracking) the buy value till it hits the trigger value, without further trailing - Does that make sense?

In a cross strategy, actual trailing (in the sense of how PT utilizes it) would always buy higher, and sell lower than using a very small simple spread as a trigger for both the buy and sell.

In other words, the objective is to buy and sell as close to the cross as possible. The only way I think that can happen is for a user defined very small spread (buy value) as the definitive trigger (A very small positive "buy value" for a buy, and a very small negative value for a sell) That should allow the buy to stay in force for the full duration of the up trend, and only sell when the "buy value" becomes slightly negative (user defined)

Something like this:

ALL_buy_strategy = EMACROSS
ALL_buy_value = 0.05

ALL_sell_strategy = EMACROSS
ALL_sell_value = -0.05

Make sense?

**Comment by taniman** [comment]
@TheBigShort111 @ITJusus 
Are you both aware that at the movent the lines are realy realy close to each other that they can cross multiple times? 

@YoJeff  already mentioned this above i think.

So the is bound to cause a lot of fakeouts. 

And the main problem.
 
The lines will cross.  the bot will buy. 
then the lines will uncross as the candle is not closed.

So people will start complaining. The bot bought but the lines never crossed. 

Please tell me how to explain this part or make it clear that the lines did INDEED cross for 10 seconds.

**Comment by ITJusus** [comment]
@TheBigShort111 
Your right, I think, I should have used the word "monitoring" rather than " trailing".  Fits.  

@taniman
Regarding the lines crossing each other multiple times...my thought on that are:

1) Only the first cross will matter as you'll then buy the coin and each successive cross back down and up the EMA line will just be the normal kind of up and down movement of the price (>1% down / >1% up) that happens in most any new entry.  Unless you happen to catch a coin in an explosive bull run, the price usually dips below your initial entry price a bit here and there for a while, until it climbs up a bit more.

2) I do see your point that you may be worried about people whining the EMA lines didn't actually cross, from a support/dev perspective, but I would think anyone making use of the "0" candle setting would except ahead of time that it is often going to look like a mesh of two EMA lines on top of each other when a buy entry is triggered and they would except that lack of obvious looking spread and just a necessary aspect of using such a strategy.  As far as the close of the candle though, the wick should still be apparent that it at least touched the line.

3) I would also agree that one problem that would be caused by this is if someone were trying to utilize a 0 candle or low spread EMACROSS strategy for the actual **sell** of the coin.  I could see that causing very rapid entries and exits with little profits...unless you have a way of instructing PT to not sell on such an EMACROSS if some minimum profit is not achieved.  I wasn't even thinking of the EMACROSS being used to sell a coin...as it seems to me you would lose most of your profit most of the time if you tried to use the EMACROSS for selling strategy.  I was only interested in the concept of EMACROSS 0 candle as slight modification to the already existing EMACROSS buy strategy.

It just seems to me that more often than not, in the coin markets (as opposed to the FX markets), getting in after the first large candle has closed out is when the actual fakeout occurs...in the sense that it has already made the major move and is now ready to dive down to normalcy before you can even make minimal profits.  More often than not, it also seems, that once the Fast EMA is crossing up the Slow EMA...the entry closest to that initial cross does not take long to become profitable, even if after a number of dips below the Slow EMA line.

**Comment by TheBigShort111** [comment]
@taniman 
I am completely aware that when the lines are close they can cross multiple times in a single candle, which is why I think the best method for validating the upward cross is to use a user defined spread for a trigger. The smaller the spread, the closer to the cross the buy or sell will trigger, with an increased risk of a reversal of the cross. The bigger the spread the more validation of the cross, with a respective reduction of overall gain.

If no cross validation method is utilized, then you are correct, the scenario you mention with multiple crosses is inevitable... However, a small to moderate spread would validate the cross in the overwhelming majority of situations - Again, the user would specify the cross spread and would be determining their own level of cross-back risk.

For example a spread of around 0.1 for a buy would validate most crosses, and a very small spread like -0.01 for a sell would work great - Reason being, the buy to have validation, and the sell to happen as close to the downward cross as possible.

Anyone using this method should be fully aware that there will be occasional instances when the cross method will result in small losses (mostly trading fees). This can happen even when an upward cross has been validated... The reason being, occasionally an upward cross happens with minimal gain, then reverses resulting in a small loss. This happens MOSTLY in sideways market movement. However, in bull runs the EMACROSS as a sell strategy would outperform the GAIN strategy in the overwhelming majority of cases, proving MUCH more lucrative.

So either using a spread to specify the buy trigger, or using the cross candle close as the trigger, either would work, however the spread method puts control in the user's hands, and would prove more valuable and predictable as it is a defined value.

The case for EMACROSS as a sell strategy versus GAIN in a bullish trend can be proven countless times on any chart. With gain, your position sells once your minimum gain has been met and any trailing has been satisfied, even if the price continues to climb before "crossing over" - In most cases that difference is very significant.

The nice thing about EMACROSS as a sell strategy is that it will ALWAYS get out of a position at the downward cross, which would prove VERY valuable when there are dumps, and selloffs like what we have been experiencing over the past couple of days. I'm holding bags that I wouldn't be holding with EMACROSS, as it would not be buying and DCA'ing all the way down... It would wait for a reversal, then buy, then if there were another selloff, it would sell at the cross, which would prove invaluable.

I sincerely appreciate your effort to bring this to fruition, and I am willing to help in any way I can to make the strategy work as efficiently as possible... I am available for discussion anytime that is convenient for you if you would like to discuss additional details. Perhaps we could video chat, or have a phone conversation?

@ITJusus 
The EMACROSS as a sell strategy works best in bullish markets as would almost always out perform the gain strategy for reason outlined in one of my previous posts. Now that a strategy can be specified per market condition in feeder, this would prove a VERY lucrative strategy.

**Comment by ITJusus** [comment]
@TheBigShort111 

I see what you're saying.  I can definitely see EMACROSS (at the 0 candle) as a great selling strategy in a bull or even gradually upward trending market with enough volume behind it.

**Comment by TheBigShort111** [comment]
@ITJusus 
Yes, absolutely...

The problem with GAIN as a sell strategy is that it's "always" a compromise. Whereas with EMACROSS it is only sometimes a compromise.

With GAIN on the sell side, these are scenarios that come into play. 1) If you set your minimum gain too high, you end up missing all of the smaller gains, and can end up with bags if you minimum gain % and trailing amount is not satisfied before a downturn in the market. 2.) If you set a smaller minimum gain and smaller trailing amount, then you miss the bulk of any bull run as the bot will sell with any dip that satisfies your trigger on the way up. 3.) If you specify a larger trailing profit then you lose some of the gain up top and pretty much all but negate any smaller gains. This method turns modest gains into slight gains, and potentially decent gains into modest ones. 4.) If you specify a more conservative trailing value you end up selling too early.

Don't get me wrong, there are instances where GAIN as a sell strategy works well, especially in a sideways market, which is why I firmly believe utilizing BOTH strategies would prove a great combination. GAIN in the sideways markets, and EMACROSS in the bull and bear markets.

Right now I have several bags that were bought right before the most recent selloff a couple of days ago... The conditions were right for the buys, but they didn't get over my sell value and trailing (which is conservative) using GAIN, not only do I have the bags, but they have since fully DCA'd and I'm still sitting on them. With EMACROSS on the sell side, these positions would have dumped at the downward cross, and I would likely not be holding any of these positions right now.

Even with a small pump on the way down, a cross would have been very unlikely during the dumps. Resulting in no buys on the way down. However, even if there was a cross up, the bot would have sold at the down cross, so I still would have all my capital free at the bottom of the market to be trading now and taking advantage of the market, instead of sitting on the bags with my capital tied up.

See my point?

**Comment by Circumvenscion** [comment]
I've been asking for this for months now. EMACROSS as a sell strategy would be unbelievably great in certain instances. The constant crossover in price action is something that would need to be addressed, but when a coin goes on a bull run, more often than not with ANY other sell strat - you miss most of those gains because of the trailing profit that gets triggered on naturally occurring swings. If we could use EMACROSS as a sell strat, then it would capture ALL of those beautiful gains, and then lose a slight amount on the back-end as the emas cross back down, but the profits would be much much greater than anything GAIN or HIGHBB could ever achieve.

**Comment by keithofaptos** [comment]
YES PLEASE +1!

**Comment by TheBigShort111** [comment]
@Circumvenscion , @keithofaptos  - From what I understand, it is definitely in the works. I've been trying to help in any way I can to bring this to fruition. Yes, EMACROSS on the sell side would prove very profitable under certain circumstances, however, by nature there will be false signals that will result in occasional small losses. So be advised that unlike the current gain method, where a coin goes into DCA till it turns profitable, EMACROSS will sell the position at the down cross, even if at a loss.

To me, and surely most that understand the benefit to using EMACROSS on the sell side, those small losses are something I'm willing to deal with for the benefit of capitalizing on the bull runs. Another HUGE benefit is the bot not buying on the way down in a bear condition, till there is a stable bottom and a reversal to trigger a buy. Again, there will be instances with false signals from time to time, but overwhelmingly this strategy would outperform GAIN in bull runs, and bear selloffs.

**Comment by taniman** [comment]
I just want to make something clear. The problem in question is not EMACROSS as a sell strategy. This issue is about something completely different. It's about how EMACROSS should behave.

EMACROSS as a sell strategy will be released with PT 2.0 so let's stop talking about it in this thread

**Comment by ITJusus** [comment]
Yes, please, while I understand the desire for EMACROSS as a sales strategy...that discussion is starting to verge on the point of hijacking this thread.  Let's please focus on the subject of "EMACROSS from candle 0" related to the Buy Strategy in this thread and keep the EMACROSS sell strategy talk over in one of the many other threads that have already discussed that topic.

**Comment by TheBigShort111** [comment]
Oh, right, sorry guys... I see your point.

So, getting back on topic, I agree completely that EMACROSS in its current form is unreliable and unpredictable as a buy strategy. However, with a couple of sensible controls it could be great. I'm a firm believer in using a small spread to define buy trigger versus using candles... Here's why...

A single candle can vary a significant amount (especially when you factor in the length of the indicator candle) - A 1 minute candle will generally have less price swing than a five minute candle, or 15 minute candle... So waiting several candles before making a buy does not make much sense.

As ITJusus has indicated, the buy needs to happen with much more predictability and as close to the cross as possible. I have spent many hours going back and manually testing my small spread theory and it proves to be a sound method for validating a cross. In addition to using a spread, there could be an "optional" time-based user-defined control.

The user defined buy timer would/could be used in conjunction to the spread to prevent the buy from happening till both conditions are met. So, the logic would be... A spread must remain positive by "at least" the user defined value and remain positive after the user defined timer period has elapsed... Which would validate and initiate the buy.

This would allow the user to set a comfort level for validating the buy which should reduce/eliminate the fluctuations back and forth at the cross candle. In both cases, a zero value should be allowed and PT would simply buy a position when the buy value reaches 0.00% - I know there is fluctuation at the cross, but we are talking only about a EMACROSS buy here, not a buy and sell scenario... So the user should have the option to set any value for both the buy value and the timer (if used)

Perhaps something like this:

ALL_buy_strategy = EMACROSS

EMA_buy_spread = 0.05
"spread at which either a buy happens instantly, OR the value at which the optional timer is triggered"

EMA_buy_timer = 30 (optional) "value set in seconds, or false"
"number of seconds to wait before initiating the buy. If the "current value" is negative after the "EMA_buy_timer" PT does nothing. If the "current value" remains positive by "at least" the spread the PT buys the position.

This optional timer function would work very similarly to using the current candle method. However, this method should prove more reliable as it gives the user definitive control over the timer, or even disable it. Disabling the timer allows the user to rely solely on the spread, which could also be set to zero... Essentially allowing a "zero value" cross scenario if desired.

Again, the timer should be optional and a zero value should be allowed for the buy spread.

Unless I'm over simplifying this, it would seem to me that the timer function could be easily adapted from the current candle method, and the basic spread function could be lifted from the current EMASPREAD strategy... In other words, it seems PT already has much of the needed ability to pull this off, with just some re-configuring... No? - In no way am I trying to mitigate the work involved, I'm just saying that (from a layman's perspective) it seems that both of these controls are not a far stretch from PT's current abilities.

That said, my sincerest thanks to the devs for their hard work, and for taking the time to consider the wants/needs of the user community. Your efforts are greatly appreciated!

**Comment by sirloinwesk** [comment]
Would it be possible to have the bot trigger the buy order based on volume amount at a particular candle after the cross? For instance if the cross happened x candles ago and volume has risen a certain number execute the buy.

Thanks for the help as always!

**Comment by ITJusus** [comment]
@sirloinwesk 

Though I like that idea, this sounds like an entirely new and separate feature request.  Can you create a separate thread/New request for that.  We're trying to keep this thread purely regarding "EMACROSS from candle 0" and not a discussion about everything EMACROSS.

**Comment by taniman** [comment]
@ITJusus I have spend the last 30 minutes investigating this question anad this is what I can see. 
Usin EMA_cross_candles = 1 is doing already the job of checking if the cross is happening RIGHT this instance.

Some proof. Here you see the cross is happening this current candle
![image](https://user-images.githubusercontent.com/6751341/37526128-e1c1af70-292e-11e8-966f-4596fcd0caf5.png)

here you see PT is ready to buy.
![image](https://user-images.githubusercontent.com/6751341/37526174-f8dc67cc-292e-11e8-90ad-116ab738becc.png)

my settings
![image](https://user-images.githubusercontent.com/6751341/37526234-1d9149de-292f-11e8-8bf1-b91313dcfbb0.png)

![image](https://user-images.githubusercontent.com/6751341/37526274-31985f08-292f-11e8-9a5e-afb50905413f.png)

So what's the problem I would ask?

**Comment by taniman** [comment]
Please take a look at our emacross indicator on trading view to see that it buys on the current candle where the cross happened if ema_cross_candles = 1

It should be live in about 30 minutes after caching clears

**Comment by ITJusus** [comment]
@taniman 

If that is the case, then I think the confusion may stem from wording in the "Buy_and_Sell_Strategies" section of the Wiki which states (for EMA_cross_candles = 2)...

"If fast EMA just crossed up the slow EMA in the last 2 candles *(exclude current candle)*"

This leads one to think that the "current candle" is always excluded.  If setting EMA_cross_candles = 2 means that you are currently on the 3rd candle after the cross and the cross actually occurred in the last 2 candles before the current candle (which seems to be the meaning of the Wiki)...then that would mean setting the EMA_cross_candles =1 would then logically translate to...

"If fast EMA just crossed up the slow EMA in the last 1 candle *(exclude current candle)*"

This is what has led us to think that we must be able to set EMA_cross_candles = 0 in order to achieve recognition of EMA cross in the current candle.  I will be very glad to know that my understanding of the Wiki is not true and EMA_cross_candles = 1 actually translates to...

"If fast EMA just crossed up the slow EMA in the *current* candle *(current candle)*"

But every time I've tested that doesn't seem to be the case.  I'll test again.  I could be mistaken.

The other problem is suppose the EMA cross does occur in the current candle and EMA_cross_candles = 1 really does mean the current candle...and then the price goes up 2% in that same candle and you just so happen to start PT at that point of your day, at which time it happens to only go up about .1% and then plummets.  This means, if say your buy_value was 0.5 you will get in on the drop.  Not ideal.   Unless it only allows a buy to occur when the Fast EMA is between the cross of the Slow EMA and the % set as the buy_value (above the cross)...but I'm under the impression that it will buy at any price when the Fast EMA is anywhere at or above the buy_value set.

Even if that is the case, another thing that could happen in this scenario is you might catch the price as the Fast EMA is about to cross back *DOWN* the Slow EMA (after it made that 2% rise) and so just because the EMA cross happened to occur at some point in the current candle and is currently in the buy_value % (0.5) above the original cross specified, it will buy anyway (blind to direction).  We would only like it to buy when it is crossing *UP* and not when it happens to be crossing back *DOWN* in the same candle as well as not when it has already rocketed way above our set buy_value.

This is why we thought it may be necessary to change the tracking behavior of the EMA Cross Strategy in general.

**Comment by taniman** [comment]
@ITJusus If you indicate a buy_value of 0.01 and the current spread is 2% is will indeed also buy.

That is why we have the buy_value_limit

so set it like this.
ALL_buy_value = 0.05 
ALL_buy_value_limit = 0.2

so the bot will only buy if the price is in that range.

we have indeed seen that the wiki is not explaining this correctly and will change this soon

**Comment by ITJusus** [comment]
@taniman 

Sorry about that, my brain was partially thinking about the HIGHBB request I have open (which can't utilize ALL_buy_value_limit)...EMA Cross does have that option.  Thanks.

Great so, now assuming us setting EMA_cross_candles = 1 actually translates to...

"If fast EMA just crossed up the slow EMA in the current candle (current candle)"

this only leaves one weakness in the strategy...

Even if that is the case, another thing that could happen in this scenario is you might catch the price as the Fast EMA is about to cross back DOWN the Slow EMA (after it finished making that 2% rise) and so just because the EMA cross happened to occur at some point in the current candle and the Fast EMA is currently again down in the buy_value % (0.5) above the original cross specified, it will buy anyway (blind to direction).  This would be buying on the way down. 

We would only like it to buy when it is crossing UP and not when it happens to be crossing back DOWN in the same candle.  I don't know how hard that would be to pull off, programmatically, but that would be the ideal.

**Comment by ITJusus** [comment]
@taniman 

One other possible weakness with EMA Cross would be, what if the price quickly jumps up through (over) our buy_value to buy_value_limit window?  Say our settings are as follows...

EMA_cross_candles = 1
ALL_buy_value = 0.05
ALL_buy_value_limit = 0.2

The Fast EMA could hypothetically move up and over that tight window in one price move (according to what PT is able to detect) and so move from being 0.02 above the Slow EMA to immediately 0.3 above the Slow EMA.  Wouldn't PT miss this buy opportunity entirely and then if the price goes up 1% in that same candle and then happens to land in that 0.02-0.3% range on the way back **DOWN**...would it not then force us to buy the coin on the way down (bag)?

Things like that are why we thought it may be necessary to change the tracking behavior of the EMA Cross Strategy, when it determines how and when to buy, in general.
